lib/Target/CBackend/CBackend.cpp

```diff
@@ -5315,26 +5330,38 @@ void CWriter::printGEPExpression(Value *Ptr, gep_type_iterator I,
   }
 
   for (; I != E; ++I) {
+    Value *Opnd = I.getOperand();
+
     cwriter_assert(
-        I.getOperand()
+        Opnd
             ->getType()
             ->isIntegerTy()); // TODO: indexing a Vector with a Vector is valid,
                               // but we don't support it here
+
     if (I.isStruct()) {
-      Out << ".field" << cast<ConstantInt>(I.getOperand())->getZExtValue();
+      Out << ".field" << cast<ConstantInt>(Opnd)->getZExtValue();
     } else if (IntoT->isArrayTy()) {
-      Out << ".array[";
-      writeOperandWithCast(I.getOperand(), Instruction::GetElementPtr);
-      Out << ']';
+      // We see through zero-element arrays and use direct pointer arithmetic
+      // instead. See: skipEmptyArrayTypes
+      if (IntoT->getArrayNumElements() == 0) {
+        if (!isConstantNull(Opnd)) {
+          Out << " + (";
+          writeOperandWithCast(Opnd, Instruction::GetElementPtr);
+          Out << ')';
```

Comment:
Ah, this is going to not go very well if we then have a non-zero-length array inside it, we'd get `(ptr) + (0).array[foo]`…

I need to add an extra layer of brackets, but the annoying thing is I'd need an extra one for each array. Now I see why the vector code only works if there's only one level of vector.

Reply:
Maybe we should print GEPs recursively from right to left.

Reply:
…though that would be very annoying considering how the GEP type iterator works.

Reply:
I have figured out how to describe the current limitations and written an `assert`. In the long term a comprehensive solution would be better, but I hope this is fine for now.

Reply:
`&(*ptr).array[n] + Opnd` should be exactly the same as `&(*ptr).array[n][Opnd]` after parsing, due to the way that `a[i]` is defined in C.

I'm not sure I fully understand how this proposed scheme should work. In particular, in the struct psuedo-definition:
```
struct annoying {
  int a[0];
  char b;
  int c[0];
};
```

The Array type for `a` and `c` seem like they may have subtle implications on pointer re-alignment which don't seem to be trivial. Perhaps that is okay (I attempt elsewhere point out that llvm-cbe is not ABI-compatible with existing C libraries), but I'm not sure the best mapping here.

Another possible strategy for GEP might be to split the expression sometimes (if we happen to know that `Value *`Ptr is an `Instruction *`):
```
_1 = &_0[n].array[n]; _1 = &(*(IterType*)((char*)_1 + n)).array[n];
```

Reply:
> `&(*ptr).array[n] + Opnd` should be exactly the same as `&(*ptr).array[n][Opnd]` after parsing, due to the way that `a[i]` is defined in C.

That doesn't sound right. `*(&(*ptr).array[n] + Opnd)` would be the same as `&(*ptr).array[n][Opnd]`, but there I have added `*` and brackets.

What I am trying to do here is to offset the pointer without going one level deeper into the structure. For example, `skipEmptyArrayTypes` would turn `[10 x [10 x i8]]*`, which has four levels (pointer, outer array, inner array, integer), into `i8*`, which has two levels. If I then encounter a GEP for that type with the indices `0, 0, 2`, I can't translate that as `&(*ptr + 0)[0][2]` because it would dereference it two times too many. However, `&(*ptr + 0 + 0 + 2)` is perfectly fine.

It would be possible to achieve the same thing with `[]` instead of `+`, but I'd have to add extra `&`'s at the front, which would be a bit annoying.

> I'm not sure I fully understand how this proposed scheme should work. In particular, in the struct psuedo-definition: […]

The scheme I'm suggesting doesn't work for zero-element arrays within structs, indeed. This pull request doesn't touch those though, so far as I know.

> Another possible strategy for GEP might be to split the expression sometimes

That's an interesting idea. I guess we'd have to disable inlining of GEP instructions in that case, though, which doesn't seem very appealing.

Reply:
(I edited my examples (`&*(ptr + 0)[0][2]` and `&*(ptr + 0) + 0 + 2`) more than once to try to make them more readable, but probably just made things worse, sorry about that.)